language_service/src/hover.rs

```diff
@@ -25,67 +29,458 @@ pub(crate) fn get_hover(
+                        self.end = ident.span.hi;
+                    } else {
+                        self.visit_ty_def(def);
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+
+    fn visit_ty_def(&mut self, def: &'_ ast::TyDef) {
+        if span_contains(def.span, self.offset) {
+            if let ast::TyDefKind::Field(ident, ty) = &*def.kind {
+                if let Some(ident) = ident {
+                    if span_contains(ident.span, self.offset) {
+                        self.header = Some(header_from_name(
+                            &ident.name,
+                            &get_type_name_from_ast_ty(ty),
+                        ));
+                        self.start = ident.span.lo;
+                        self.end = ident.span.hi;
+                    } else {
+                        self.visit_ty(ty);
+                    }
+                } else {
+                    self.visit_ty(ty);
+                }
+            } else {
+                walk_ty_def(self, def);
+            }
+        }
+    }
+
+    fn visit_pat(&mut self, pat: &'_ ast::Pat) {
+        if span_contains(pat.span, self.offset) {
+            match &*pat.kind {
+                ast::PatKind::Bind(ident, anno) => {
+                    if span_contains(ident.span, self.offset) {
+                        self.header =
+                            Some(header_from_name(&ident.name, &self.get_type_name(pat.id)));
+                        self.start = ident.span.lo;
+                        self.end = ident.span.hi;
+                    } else if let Some(ty) = anno {
+                        self.visit_ty(ty);
+                    }
+                }
+                _ => walk_pat(self, pat),
+            }
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &'_ ast::Expr) {
+        if span_contains(expr.span, self.offset) {
+            match &*expr.kind {
+                ast::ExprKind::Field(_, field) if span_contains(field.span, self.offset) => {
+                    self.header = Some(header_from_name(&field.name, &self.get_type_name(expr.id)));
+                    self.start = field.span.lo;
+                    self.end = field.span.hi;
+                }
+                _ => walk_expr(self, expr),
+            }
+        }
+    }
+
+    fn visit_path(&mut self, path: &'_ ast::Path) {
+        if span_contains(path.span, self.offset) {
+            let res = self
+                .compilation
+                .unit
+                .ast
+                .names
+                .get(path.id)
+                .unwrap_or_else(|| panic!("Can't find definition for reference node: {}", path.id));
+            match &res {
+                resolve::Res::Item(item_id) => {
+                    let item = find_item(self.compilation, item_id).unwrap_or_else(|| {
+                        panic!("Can't find definition for reference node: {}", path.id)
```

Comment:
This should return a `None` hover instead of panicking

Reply:
I've removed the panics that could result from incomplete Q# code.